capstone/files/tests/test_main.py

```diff
@@ -0,0 +1,65 @@
+from fastapi.testclient import TestClient
+from fastapi import UploadFile
+
+from os import path
+
+import builtins
+import main
+import tempfile
+import pytest
+
+client = TestClient(main.app)
+
+
+def test_homepage():
+    response = client.get('/')
+    assert response.status_code == 200
+    assert response.template.name == "form.html.jinja"
+    assert "request" in response.context
+
+
+def test_download_file_ok(monkeypatch):
+    f = tempfile.NamedTemporaryFile()
+    content = b"This is an example \x00\xff"
+    f.write(content)
+    f.seek(0)
+
+    def join(a, b):
+        return f.name
+
+    monkeypatch.setattr(path, "join", join)
+
+    response = client.get('/1111')
+
+    assert response.content == content
+    f.close()
+
+
+@pytest.mark.parametrize("file_name", [
+    "some-file", "../main.py"
+])
+def test_download_file_bad(file_name):
+    response = client.get(file_name)
+    assert response.status_code == 400
+
+
+def test_upload_file(monkeypatch):
+    def open(path, type):
+        return tempfile.TemporaryFile()
```

Comment:
No need to use tempfile, you can post file content as string

Reply:
shutil.copyfile needs an actual file as a destination.
I can mock it too, but why don't we just use a tempfile for that?

Reply:
For `shutil` you're already opening a file, so this tempfile has nothing to do with it. You can do the test request as:

```python
    response = client.post("/upload-file/",
                           files={
                               "file": ('testname.txt', b'This is an example \x00\xff')
                           })
```

In the server side, you can use unittest's [`mock_open`](https://docs.python.org/3/library/unittest.mock.html#mock-open) to avoid writing an actual file to disk.